org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/preferences/internal/DefaultPersistentModelBuilder.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2017 the original author or authors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.buildship.core.preferences.internal;
+
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.jdt.core.IClasspathEntry;
+
+import org.eclipse.buildship.core.preferences.PersistentModel;
+import org.eclipse.buildship.core.preferences.PersistentModelBuilder;
+
+/**
+ * Default implementation for {@link PersistentModelBuilder}.
+ *
+ * @author Donat Csikos
+ */
+public final class DefaultPersistentModelBuilder implements PersistentModelBuilder {
+
+    private static final IPath DEFAULT_BUILD_DIR = new Path("build");
+    private static final Collection<IPath> DEFAULT_SUBPROJECTS = ImmutableList.<IPath>of();
+    private static final List<IClasspathEntry> DEFAULT_CLASSPATH = ImmutableList.<IClasspathEntry>of();
+    private static final Collection<IPath> DEFAULT_DERIVED_RESOURCES = ImmutableList.<IPath>of();
+    private static final Collection<IPath> DEFAULT_LINKED_RESOURCES = ImmutableList.<IPath>of();
+
+    private final IProject project;
+
+    private IPath buildDir = DEFAULT_BUILD_DIR;
```

Comment:
There should be no defaults. The builder should fail hard if something isn't specified.